### PR TITLE
Add GH Workflow/precommit check to `fmt`, `vet`, and `mod tidy`.

### PR DESCRIPTION
**Pull request by kaovilai**
Using github actions (outside openshift org on forks) to notify dev of fmt, vet, or mod issues.

See run https://github.com/kaovilai/oadp-operator/actions/runs/3206089810

Bonus: optional precommit to block untidied, unformatted commits. Manual install required.

### CODE REVIEW

**Comment by sseago** [comment]
I guess the main risk of bot commits to the PR is they get overwritten with force pushes. I'm assuming that a force push will just regenerate the bot commits, so it may be fine. The alternative (which is what happens upstream with Velero) is a ci task that fails with an error message asking the user to tidy/format.

**Comment by kaovilai** [comment]
Force push regenerates bot commits as long as github actions isn't offline in your fork. We cannot use github actions in this org so it won't show up in statuses here. This is run from the fork side only.

**Comment by kaovilai** [comment]
But yes can make the action fail from the fork side.

**Comment by kaovilai** [comment]
Failing in a CI task requires you to push again at a later time. Pushing format and tidy for you saves dev time. They can just pull the tidy.. or force push a tidied commit if they wish.

Any accidental untidied force pushes will be auto fixed in the next bot commit.

**Comment by shawn-hurley** [comment]
This would, at least for me, make my life much more challenging in managing my branches. Having something auto-commit w/o the end user asking for it scares me.

**Comment by sseago** [comment]
@shawn-hurley I have similar concerns. I'm not in the habit of fetching my PR branches after I create them. In fact on a recent PR after I merged a commit to my PR branch from someone else, I subsequently forgot that I did that and force-pushed a rebase change, which lost those remote PR changes.

**Comment by shawn-hurley** [comment]
It is the same reason that I don't like to use the commit suggestions. I always get divergent branches, I can recover it but just something to consider.

**Review by weshayutin — COMMENTED** [review]
checking this out now .. I like the intent for sure

**Comment by weshayutin** [comment]
so in pelorus we have a precommit hook that works locally for developers that run any lint and format prior to the developer pushing the code.  Not sure if I 100% understand this yet, but I'd rather not have code fixed in CI and committed by ci tooling.
Prevent the commit locally if lint fails.

https://github.com/konveyor/pelorus/blob/master/Makefile#L75
https://github.com/konveyor/pelorus/blob/master/scripts/pre-commit

not 100% sure if this applies here yet.. but look at it more.

File: .github/workflows/go-autoformat.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    - name: git config
+      run: |
+        git config --global user.email "autoformat@oadp.io"
+        git config --global user.name "Auto Format"
+    - name: format
+      run: go fmt ./...
+    - name: commit & push changes # push step 1, show missing format before vet and tidy starts to provide early feedback
+      run: |
+        git add .
+        git commit -m "Auto Format: 'go fmt ./...'" && git push
```
Comment (weshayutin) [inline .github/workflows/go-autoformat.yml:38]: ya.. not ok.. imho.   
enforce the linting, code spec locally before it leaves a developer's laptop :)

**Comment by kaovilai** [comment]
@weshayutin @shawn-hurley @sseago addressed feedback. doing `git diff --exit-code` instead of committing.

**Comment by kaovilai** [comment]
Updated PR description. See run example here. https://github.com/kaovilai/oadp-operator/actions/runs/3206089810

PS. 4.10 flake
/retest

**Review by sseago — APPROVED** [review]

**Comment by kaovilai** [comment]
[4.9 succeeded but failed ci deprovisioning](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_oadp-operator/839/pull-ci-openshift-oadp-operator-master-4.9-operator-e2e-aws/1578415642103517184#1:build-log.txt%3A55-56).
/override 4.9-operator-e2e-aws

[4.11 succeeded but failed ci must-gather](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_oadp-operator/839/pull-ci-openshift-oadp-operator-master-4.11-operator-e2e-aws/1578415633224175616#1:build-log.txt%3A55-56)
/override 4.11-operator-e2e-aws

**Comment by kaovilai** [comment]
[4.9 succeeded but failed ci deprovisioning](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_oadp-operator/839/pull-ci-openshift-oadp-operator-master-4.9-operator-e2e-aws/1578415642103517184#1:build-log.txt%3A55-56).
/override ci/prow/4.9-operator-e2e-aws

[4.11 succeeded but failed ci must-gather](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_oadp-operator/839/pull-ci-openshift-oadp-operator-master-4.11-operator-e2e-aws/1578415633224175616#1:build-log.txt%3A55-56)
/override ci/prow/4.11-operator-e2e-aws

**Comment by kaovilai** [comment]
/test ci/prow/4.10-operator-e2e-aws
[`level=error msg=Error: Provider produced inconsistent result after apply`](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_oadp-operator/839/pull-ci-openshift-oadp-operator-master-4.10-operator-e2e-aws/1578458080566317056#1:build-log.txt%3A64)

**Comment by kaovilai** [comment]
That moment when the test and override context are different :/
/test 4.10-operator-e2e-aws
[level=error msg=Error: Provider produced inconsistent result after apply](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_oadp-operator/839/pull-ci-openshift-oadp-operator-master-4.10-operator-e2e-aws/1578458080566317056#1:build-log.txt%3A64)

**Comment by kaovilai** [comment]
Adding a precommit

File: Makefile [hunk 2]
```diff
@@ -429,3 +432,16 @@ test-e2e: test-e2e-setup ## execute the oadp integration tests
...
+precommit-git-hooks-install: ## Install git hooks for precommit
+	@echo "Installing git hooks"
+	@echo '#!/bin/sh' > .git/hooks/pre-commit
+	@echo 'make precommit' >> .git/hooks/pre-commit
+	@echo 'git diff --exit-code go.mod go.sum &>/dev/null' >> .git/hooks/pre-commit
+	@echo 'if [ $$? -ne 0 ]; then' >> .git/hooks/pre-commit
+	@echo '  echo "go.mod and go.sum are out of sync. Please add it to your commit"' >> .git/hooks/pre-commit
+	@echo '  exit 1' >> .git/hooks/pre-commit
+	@echo 'fi' >> .git/hooks/pre-commit
+	@chmod +x .git/hooks/pre-commit
```
Comment (kaovilai) [inline Makefile:447]: Now if you run `make precommit-git-hooks-install` you will have tidy run for you before you submit.

If your go.mod and go.sum now differs you will be prompted to add it to the next commit.
@weshayutin

**Pull request closed without merging**